Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -6709,58 +6713,84 @@ void WebPageProxy::startNetworkRequestsForPageLoadTiming(WebCore::FrameIdentifie
+
     if (!m_pageLoadTiming)
         return;
     m_pageLoadTiming->updateEndOfNetworkRequests(timestamp);
     if (m_framesWithSubresourceLoadingForPageLoadTiming.isEmpty())
-        generatePageLoadingTimingSoon();
+        generatePageLoadTimingSoon();
 }
 
-void WebPageProxy::generatePageLoadingTimingSoon()
+void WebPageProxy::generatePageLoadTimingSoon()
 {
+#if PLATFORM(COCOA)
+    static bool shouldLog = CFPreferencesGetAppBooleanValue(CFSTR("WebKitDebugGeneratePageLoadTimingLogs"), kCFPreferencesCurrentApplication, nullptr);
+
+    auto result = generatePageLoadTimingSoonImpl();
+    if (shouldLog)
+        WEBPAGEPROXY_RELEASE_LOG(Loading, "generatePageLoadTimingSoon: %d", static_cast<int>(result));
+#else
+    generatePageLoadTimingSoonImpl();
+#endif
+}
+
+WebPageProxy::GeneratePageLoadTimingResult WebPageProxy::generatePageLoadTimingSoonImpl()
+{
+    using enum WebPageProxy::GeneratePageLoadTimingResult;
+
     m_generatePageLoadTimingTimer.stop();
     if (!m_pageLoadTiming)
-        return;
-    if (m_framesWithSubresourceLoadingForPageLoadTiming.size())
-        return;
+        return WaitForPageLoadTimingObject;
 
     WallTime lastTimestamp;
     if (!m_pageLoadTiming->firstVisualLayout())
-        return;
+        return WaitForFirstVisualLayout;
     lastTimestamp = std::max(lastTimestamp, m_pageLoadTiming->firstVisualLayout());
 
     if (!m_pageLoadTiming->firstMeaningfulPaint())
-        return;
+        return WaitForFirstMeaningfulPaint;
     lastTimestamp = std::max(lastTimestamp, m_pageLoadTiming->firstMeaningfulPaint());
 
     if (!m_pageLoadTiming->documentFinishedLoading())
-        return;
+        return WaitForDOMContentLoaded;
     lastTimestamp = std::max(lastTimestamp, m_pageLoadTiming->documentFinishedLoading());
 
     if (!m_pageLoadTiming->finishedLoading())
-        return;
+        return WaitForLoadEvent;
     lastTimestamp = std::max(lastTimestamp, m_pageLoadTiming->finishedLoading());
 
-    if (!m_pageLoadTiming->allSubresourcesFinishedLoading())
-        return;
+    if (m_framesWithSubresourceLoadingForPageLoadTiming.size())
+        return WaitForSubresourcesFinishedLoading;
     lastTimestamp = std::max(lastTimestamp, m_pageLoadTiming->allSubresourcesFinishedLoading());
 
     // Stop waiting for page load to end 100 ms after the last of the timestamps we care about.
     Seconds interval = std::max(0_ms, lastTimestamp + 100_ms - WallTime::now());
     m_generatePageLoadTimingTimer.startOneShot(interval);
+    return WaitForQuiescence;
 }
 
 void WebPageProxy::didEndNetworkRequestsForPageLoadTimingTimerFired()
 {
     if (!m_pageLoadTiming)
         return;
 
+    // If the page had no subresources, just say that all subresources finished loading when the
+    // load event fired.
+    if (!m_pageLoadTiming->allSubresourcesFinishedLoading()) {
+        auto loadEventEnd = m_pageLoadTiming->finishedLoading();
```

Comment:
For some reason when I wrote this as a one line statement I got the "misleading indentation" warning.